Enable text wrapping inside a vertical sizer for wxStaticText

RobertRoeb opened the pull request:
Changed the wxWrapSizer code from using a variable "m_usedLast" to two different functions
  CalcMinFirstPass() and CalcMin() for the two step min size calculation.
Added the same logic and naming to controls using GetEffectiveMinSizeFirstPass() 
Implemented this for the generic code in wxStaticTextBase.
Added test to wrapsizer sample.

Commit:
Enable text wrapping inside a vertical size for wxStaticText similar to wxWrapSizer

Changed the wxWrapSizer code from using a variable "m_usedLast" to two different functions
  CalcMinFirstPass() and CalcMin() for two step min size calculation.
Added the same logic and naming to controls using GetEffectiveMinSizeFirstPass()
Implemented this for the generic code in wxStaticText
Added test to wrapsizer sample

vadz commented:
Before looking at it, could you please explain the relationship between this and `wxWrapSizer`? I mean, they do conceptually similar things, but other than that, there is none, right?

Also, I don't know if we want to do this unconditionally or if some flag is specified. While people probably don't want their labels being truncated, they may not want them to take more than one line too... And the interaction of this with ellipsize flags is not clear at all.

vadz commented on include/wx/sizer.h:
```diff
@@ -669,6 +669,9 @@ class WXDLLIMPEXP_CORE wxSizer: public wxObject, public wxClientDataContainer
     // this size to really update all the sizer items.
     virtual wxSize CalcMin() = 0;
 
+    // Can be overidden for two step process, e.g. by wxWrapSizer
+    virtual wxSize CalcMinFirstPass() { return CalcMin(); }
```
It would be nice to explain what the two step process is, because this is really not obvious.

Also, `wxWrapSizer` doesn't really _need_ it, as it worked before adding it, so this makes things even more confusing.

vadz replied:
`wxWrapSizer` was, AFAIR, somewhat of a hack and it's great to replace it with something better but this something should be documented/explained.

Maybe documenting it would also give some idea for a better name because, frankly, "first pass" is not great: it doesn't mean anything per se and implies the existence of "second pass" which doesn't actually exist.

vadz commented on include/wx/window.h:
```diff
@@ -411,6 +411,7 @@ class WXDLLIMPEXP_CORE wxWindowBase : public wxEvtHandler
         // minimum size, giving priority to the min size components, and
         // returns the results.
     virtual wxSize GetEffectiveMinSize() const;
+    virtual wxSize GetEffectiveMinSizeFirstPass() const { return  GetEffectiveMinSize(); }
```
This should really be documented (in a comment here and in `interface/wx/window.h`) because I just have no idea how to use it right now.

vadz commented on include/wx/stattext.h:
```diff
@@ -72,6 +77,12 @@ class WXDLLIMPEXP_CORE wxStaticTextBase : public wxControl
 
+    // save unwrapped label to allow to call Wrap() several times and 
+    // always starting from the original, unwrapped label
+    wxString m_unwrappedLabel;
```
Is this different from `wxControl::m_labelOrig`?

RobertRoeb replied:
I could not figure out of the stripping of markup and/or mnemonics and/or the ellipsis code would alter m_labelOrig. I think m_labelOrig is the before any change, and m_unwrappedLabel is after the other changes, but before wrapping.

vadz commented on include/wx/stattext.h:
```diff
@@ -72,6 +77,12 @@ class WXDLLIMPEXP_CORE wxStaticTextBase : public wxControl
+    // save unwrapped label to allow to call Wrap() several times and 
+    // always starting from the original, unwrapped label
+    wxString m_unwrappedLabel;
+    int m_currentWrap = 0;
```
What is the value of this? Wrap width? With `0` meaning "none"? Please comment it too.

vadz commented on samples/wrapsizer/wrapsizer.cpp:
```diff
@@ -118,10 +118,13 @@ WrapSizerFrame::WrapSizerFrame()
 
         sizerMidWrap->Add(chk, wxSizerFlags().Centre().Border());
     }
-
```
Please avoid mixing whitespace-only changes with significant ones if possible.

vadz commented on samples/wrapsizer/wrapsizer.cpp:
```diff
@@ -118,10 +118,13 @@ WrapSizerFrame::WrapSizerFrame()
     sizerMid->Add(sizerMidWrap, wxSizerFlags(100).Expand());
     sizerRoot->Add(sizerMid, wxSizerFlags(100).Expand().Border());
 
+    // A long wxStaticText that wraps like a wxWrapSizer
```
Really not sure this belongs to this sample anyhow. I'd rather add a button setting a long label to the "widgets" sample.

vadz commented on src/common/stattextcmn.cpp:
```diff
@@ -213,8 +213,46 @@ class wxLabelWrapper : public wxTextWrapper
 
 void wxStaticTextBase::Wrap(int width)
 {
+    if (width == m_currentWrap) return;
```
Please don't do this:

```suggestion
    if (width == m_currentWrap)
        return;
```

vadz commented on src/common/stattextcmn.cpp:
```diff
@@ -213,8 +213,46 @@ class wxLabelWrapper : public wxTextWrapper
+    m_currentWrap = width;
+
+    // Allow for repeated calls to Wrap with different values
+    if (m_unwrappedLabel.IsNull()) 
```
Please use standard functions instead of wx 1.x ones

```suggestion
    if (m_unwrappedLabel.empty()) 
```

vadz commented on src/common/stattextcmn.cpp:
```diff
@@ -213,8 +213,46 @@ class wxLabelWrapper : public wxTextWrapper
+    // While wxWrapSizer can only wrap entire controls, a text paragraph
+    // could theoretically wrap at a few letters, so we start with
+    // requesting very little space in the first pass
+    return wxSize( 10, 10 );
```
Could we avoid hardcoding magic numbers please? It would make sense to use `GetCharWidth()` or something else based it.

vadz commented on src/common/stattextcmn.cpp:
```diff
@@ -213,8 +213,46 @@ class wxLabelWrapper : public wxTextWrapper
+    // space in the horizontal direction. Wrap there and report a new
+    // GetEffectiveMinSize() from then on.
+
+    if (direction != wxHORIZONTAL) return false;
```
```suggestion
    if (direction != wxHORIZONTAL)
        return false;
```

vadz commented on src/common/stattextcmn.cpp:
```diff
@@ -213,8 +213,46 @@ class wxLabelWrapper : public wxTextWrapper
+    if (direction != wxHORIZONTAL) return false;
+
+    int style = GetWindowStyleFlag();
+    SetWindowStyleFlag( style | wxST_NO_AUTORESIZE );
```
It looks like we could avoid calling it unnecessarily (and also below):

```suggestion
    if ( !(style & wxST_NO_AUTORESIZE) )
        SetWindowStyleFlag( style | wxST_NO_AUTORESIZE );
```

vadz commented on src/common/wincmn.cpp:
```diff
@@ -859,21 +859,39 @@ wxSize wxWindowBase::GetEffectiveMinSize() const
     // merge the best size with the min size, giving priority to the min size
     wxSize min = GetMinSize();
 
+    const wxStaticText *text = dynamic_cast<const wxStaticText*>(this);
```
Please remove this and all the other changes to this file.

vadz reviewed (commented):
Sorry, there are a number of things to fix here, but more importantly I see 2 changes here that should ideally correspond to 2 commits:

1. Addition of `GetEffectiveMinSizeFirstPass()` and changing `wxWrapSizer` to use it.
2. Changes to `wxStaticText`.

(1) is probably fine, but it really needs to be documented.

For (2) I'm still not sure if we want to do it unconditionally or depending on a flag. If nothing else, this almost certainly shouldn't override explicit calls to `Wrap()` in the user code as it, I believe, does now.

RobertRoeb committed:
Comply to formatting

Co-authored-by: VZ <vz-github@zeitlins.org>

RobertRoeb committed:
use String

Co-authored-by: VZ <vz-github@zeitlins.org>

RobertRoeb committed:
Avoid unnecessary call of setting flag

Co-authored-by: VZ <vz-github@zeitlins.org>

Commit:
Added documentation for new xxx FirstPass() methods

  Use GetCharHeight() and GetCharWidth() as minimal size for wrapping wxStaticText
  Removed accidentally committed debug code
  Added clarifying comments to new wrapping code in wxStaticText

Commit:
Document the new behaviour in wxStaticText

vadz commented:
> This only happens inside of a vertical sizer, which calls InformsFirstDirection() to for the sole purpose of wrapping the text.

Wait, does this mean that wrapping still won't work in a `wxFlexGridSizer`? This makes it significantly less useful and potentially very, very confusing: there are no other examples of controls (AFAIK) that appear differently depending on the sizer they're placed in.

Can we make it work for 2D sizers too?

Commit:
Added missing override statement

vadz commented:
Could you please check the CI failures? There are some trivial to fix compile errors, but this also broke `wxWrapSizer` unit tests.

RobertRoeb commented:
I will look at the compile errors, but doing further test scenarios I encountered a situation that might constitute an unexpected change of behaviour. If you put a vertical wxStaticBoxSizer in a dialog and you add two wxStaticText elements into it, the previous code stretched the wxStaticBoxSizer horizontally until the wxStaticText would fit it using a single line each. Previously, there was no way to make the wxStaticText wrap elegantly if the text was too long - you could just set a hard wrap at a e.g. 200px. But the point is: if your dialog layout relied on an expectation that the text would force the owning sizer to stretch, then that layout will be broken with the new code. Here is my artistic explanation
Before
+--wxStaticBox--------------------------
|  This is a long text that stretches the box
|  This is short text
| (  )  wxCheckBox 
| What ever else here
+----------------------------------------

Now
+--wxStaticBox----------
|  This is a long text
|  that wraps in the box
|  This is short text
| (  )  wxCheckBox 
| What ever else here
+------------------------

I need to think if we need a wxST_WRAP_IN_SIZER or a wxWrapStaticText in order not to break anything.

Commit:
Added wxST_WRAP flag

  This make it possible to choose between the three wrap modes:
     - Don't wrap at all (and force owning size to stretch out
       to accomodate whole length on a single line)
     - Fixed wrapping when falling wxStaticText::Wrap()
     - Dynamic wrapping, typically with very long text
  Added test for all three modes side by side in the wrapsizer
     sample
  Documented new style flag

Commit:
Corrected indentation

Commit:
Converted tests to use new CalcMinFirstPass()

Commit:
Spelling error corrected

Commit:
Fix spelling and trailing white spaces

Commit:
Incorporated PR #751 to work with zero width spaces

 Also added test for this in wrapsizer sample

Commit:
Replaced the code with one that works in UTF8 mode

  std::string::find_last_of does not work with complex
  characters in UTF8 mode AFAIK

vadz committed:
Merge remote-tracking branch 'origin/master' into stattext-wrap

vadz committed:
Update comments in wxWrapSizer code to make more sense

They were semantically broken by the recent changes.

vadz committed:
Improve documentation of the new "first pass" functions

Try to make it more clear and correct various typos.

vadz committed:
Make wxWindow::GetEffectiveMinSize() code a bit more clear

Use wxSize::SetDefaults() instead of reimplementing it to make it more
obvious what happens here.

No real changes.

vadz committed:
Rationalize the new added "first pass" functions

Rename them to have a name actually meaning something to people not
familiar with the details of the wxSizer layout algorithm who don't even
know that it makes 2 passes.

Also call the new wxWindow function GetMinSizeUsingLayoutDirection(),
without the "Effective" part, as GetEffectiveMinSize() is supposed to be
just a trivial wrapper around GetMinSize() and should have never been
virtual and overridable in the first place, even if it was unfortunately
made so back in a20a357d92 (Make GetEffectiveMinSize virtual as not
being able to override the method, which the sizer system depends on, is
inflexible, 2009-01-30).

vadz committed:
Don't make "Static" page in the widgets sample so wide

Use multiple lines for the long label to prevent it from making the
contents of the page too wide.

Do make the static box take all available space, if there is any.

vadz committed:
Disable "Ellipsize" checkbox on the "static" widgets sample page

The text always fits initially, as the page size is computed using it,
so choosing to ellipsize it has no effect.

vadz committed:
Test wxST_WRAP in the "Static" page of the widgets sample too

Add a checkbox to allow turning on wxST_WRAP.

vadz committed:
Rename variable with a nonsensical name in the widgets sample

No real changes, just don't use "dummy text" which doesn't mean anything
even though it has been present since 39bc0347fd (added support for
ellipsization and markup in wxStaticText (modified patch 1629946),
2007-04-01) and call the flags used for plain text and text with markup
just "flagsText" and "flagsMarkupText" respectively.

This makes the code a bit more understandable.

vadz committed:
fixup! Enable text wrapping inside a vertical size for wxStaticText similar to wxWrapSizer

vadz committed:
Return correct min size for wrapped wxStaticText

Don't return ridiculously small width and, even more importantly, don't
return single line height when a wrapped label typically has more than
one line.

vadz committed:
fixup! Added documentation for new xxx FirstPass() methods

vadz committed:
fixup! Added wxST_WRAP flag

vadz committed:
fixup! Enable text wrapping inside a vertical size for wxStaticText similar to wxWrapSizer

vadz committed:
fixup! Enable text wrapping inside a vertical size for wxStaticText similar to wxWrapSizer

vadz committed:
Revert support for ZWSP in wxTextWrapper

This reverts 9d7c647f00 (Incorporated PR #751 to work with zero width
spaces, 2025-09-23) and 6f03759268 (Replaced the code with one that
works in UTF8 mode, 2025-09-23) before reimplementing this in a better
and simpler way.

vadz committed:
Add support for wrapping on non-ASCII space characters

Break lines on zero width space and other Unicode characters with the
"white space" property, but not non-breaking ones.

Closes #751.

RobertRoeb commented:
Hi again, what is the current plan on this PR? I read your comments that not everything is perfect, but I always get nervous if a PR is waiting in queue. Further improvements could also be in later PR and the current PR does solve the issue it is supposed to solve (and the code was actually planning to solve 15 years ago).

vadz committed:
Merge branch 'master' into stattext-wrap

vadz committed:
Remove unnecessary call to InvalidateBestSize()

This is already done by Wrap(), no need to redo it here.

vadz committed:
Don't bother restoring wxStaticText flags if we didn't change them

Only call SetWindowStyleFlag() to restore the flags if they had been
changed in the first place.

No real changes.

vadz committed:
Replace InformFirstDirection() with better API

The functions InformFirstDirection() in wxWindow and wxSizer forced the
code overriding it to be written in unnatural way, by storing the
parameters passed to this function and then using them in other
functions later.

Add new functions wxWindow::GetMinSizeFromKnownDirection() and
wxSizer::CalcMinSizeFromKnownDirection() taking the same parameters as
InformFirstDirection() but which can use them directly to return the
desired result.

Change wxStaticText and wxWrapSizer to use them instead of the old
function to simplify their code (that of wxWrapSizer could probably be
still simplified further, as it doesn't seem necessary to store these
values at all any longer, but for now keep it as is).

vadz commented:
I've pushed my changes to the API (see b2e20aabb1b0adcb195e1e63c607087b1634d69b), please let me know if you see any problems with them. If not, then I think this could be merged, even if I'm still almost certain that `wxWrapSizer` code is more complicated than necessary and could be drastically simplified now — but this is something we can leave for later/never...

vadz committed:
Make it more visually clear where static text should wrap

Create a window of 150px width to check that the text really wraps at
150px.

vadz committed:
Fix unwrapping wxStaticText label when wxST_WRAP is reset

This needs to be done in overridden SetWindowStyleFlag() instead of
doing it in GetMinSizeFromKnownDirection() as it could be called after a
manual call to Wrap(), which shouldn't result in unwrapping.

RobertRoeb commented:
Thanks, it works here in my tests.

tmodes-dev commented:
The new code introduce a regression in wxPropertyGrid, see comment in issue https://github.com/wxWidgets/wxWidgets/issues/25795#issuecomment-3453054879
